src/StreamJsonRpc.Tests/JsonRpcWithFatalExceptionsTests.cs

```diff
@@ -0,0 +1,321 @@
+#if !NET452
+            Assert.Equal(cts.Token, ex.CancellationToken);
+#endif
+        }
+
+        // Assert MessageHandler and Stream objects are not disposed
+        Assert.False(((DisposingMessageHander)this.clientRpc.MessageHandler).IsDisposed);
+        Assert.False(this.serverStream.IsDisposed);
+        Assert.False(this.clientStream.IsDisposed);
+    }
+
+    [Fact]
+    public async Task CancelMayStillReturnErrorFromServer()
+    {
+        using (var cts = new CancellationTokenSource())
+        {
+            var invokeTask = this.clientRpc.InvokeWithCancellationAsync<string>(nameof(Server.AsyncMethodFaultsAfterCancellation), new[] { "a" }, cts.Token);
+            await this.server.ServerMethodReached.WaitAsync(this.TimeoutToken);
+            cts.Cancel();
+            this.server.AllowServerMethodToReturn.Set();
+
+            await Assert.ThrowsAsync<TaskCanceledException>(() => invokeTask);
+        }
+
+        Assert.Equal(FaultException.Message, Server.ThrowAfterCancellationMessage);
+
+        // Assert that the JsonRpc, MessageHandler, and Stream objects are disposed after exception
+        await Assert.ThrowsAsync<ObjectDisposedException>(() => this.clientRpc.InvokeAsync(nameof(Server.ServerMethod), "testing"));
+        Assert.True(((DisposingMessageHander)this.clientRpc.MessageHandler).IsDisposed);
+        Assert.True(this.serverStream.IsDisposed);
+        Assert.True(this.clientStream.IsDisposed);
+    }
+
+    public class Server
+    {
+        internal const string ThrowAfterCancellationMessage = "Throw after cancellation";
+
+        public bool DelayAsyncMethodWithCancellation { get; set; }
+
+        public AsyncAutoResetEvent ServerMethodReached { get; } = new AsyncAutoResetEvent();
+
+        public AsyncAutoResetEvent AllowServerMethodToReturn { get; } = new AsyncAutoResetEvent();
+
+        public static string ServerMethod(string argument)
+        {
+            return argument + "!";
+        }
+
+        public Task ServerMethodThatReturnsCancelledTask()
+        {
+            var tcs = new TaskCompletionSource<object>();
+            tcs.SetCanceled();
+            return tcs.Task;
+        }
+
+        public void MethodThatThrowsUnauthorizedAccessException(string message)
+        {
+            throw new UnauthorizedAccessException(message);
+        }
+
+        public async Task AsyncMethodThatThrowsAfterYield(string message)
+        {
+            await Task.Yield();
+            throw new Exception(message);
+        }
+
+        public async Task<string> AsyncMethodThatReturnsStringAndThrows(string message)
+        {
+            await Task.Run(() => throw new Exception(message));
+
+            return "never will return";
+        }
+
+        public async Task AsyncMethodThatThrowsBeforeYield(string message)
+        {
+            throw new Exception(message);
+            await Task.Yield();
```

Comment:
If you get build warnings from dead code on this line, you can replace this method with:

```csharp
public Task AsyncMethodThatThrowsBeforeYield(string message)
{
  var tcs = new TaskCompletionSource<object>();
  tcs.SetException(new Exception(message));
  return tcs.Task;
}
```

Reply:
Awesome, thanks!